README.md
# apollo-link-batch-alias
Batches multiple operations into a single HTTP request. Uses aliasies under the hood.

## Installation

`npm install @mir-dit/apollo-link-batch-alias --save`

## Usage

```typescript
import { AliasBatchHttpLink } from '@mir-dit/apollo-link-batch-alias';

const link: AliasBatchHttpLink = new AliasBatchHttpLink({ uri, batchMax: 20, batchInterval: 300 });
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 2.0.2 (2021-09-28)

### 2.0.1 (2021-09-28)


### Bug Fixes

* type fixed ([9acdf6b](https://github.com/mir-dit/apollo-link-batch-alias/commit/9acdf6b2d8450afdd211512d5b5135b56cc39604))

## 2.0.0 (2021-09-28)


### ⚠ BREAKING CHANGES

* update on new apollo version

### Features

* update on new apollo version ([a9dabd4](https://github.com/mir-dit/apollo-link-batch-alias/commit/a9dabd42257eb59842e0c2c3477d7a22d8a0285e))

### 1.1.3 (2021-09-23)

### 1.1.2 (2021-09-22)

### 1.1.1 (2021-09-22)

## 1.1.0 (2021-09-22)


### Features

* Initial commit ([6d5da10](https://github.com/mir-dit/apollo-link-batch-alias/commit/6d5da10cc01826dd35ef0a8b2fff67d29cfd0dd9))


src/main.spec.ts
import { Observable } from 'zen-observable-ts';
import { AliasBatchHttpLink } from './main';

jest.mock('@apollo/client/link/batch', () => ({
  BatchLink: jest.fn(),
}));

window.fetch = jest.fn();

describe('AliasBatchHttpLink', () => {
  it('does not need any constructor arguments', () => {
    expect(() => new AliasBatchHttpLink()).not.toThrow();
  });

  it('should pass batchInterval, batchDebounce, batchMax, and batchKey to BatchLink', () => {
    const BatchLink = require('@apollo/client/link/batch').BatchLink;
    const LocalScopedLink = require('./main').AliasBatchHttpLink;

    const batchKey = () => 'hi';
    const batchHandler = () => Observable.of();

    new LocalScopedLink({
      batchInterval: 20,
      batchDebounce: true,
      batchMax: 20,
      batchKey,
      batchHandler,
    });

    const {
      batchInterval,
      batchDebounce,
      batchMax,
      batchKey: batchKeyArg,
    } = BatchLink.mock.calls[0][0];

    expect(batchInterval).toBe(20);
    expect(batchDebounce).toBe(true);
    expect(batchMax).toBe(20);
    expect(batchKeyArg()).toEqual(batchKey());
  });
});


jest.config.ts
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "C:\\Users\\infod\\AppData\\Local\\Temp\\jest",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  // coverageDirectory: undefined,

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "\\\\node_modules\\\\"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: 'v8',

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "json",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: 'jsdom',

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "\\\\node_modules\\\\"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jest-circus/runner",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "\\\\node_modules\\\\",
  //   "\\.pnp\\.[^\\\\]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


rollup.config.js
import resolve from '@rollup/plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import typescript from 'rollup-plugin-typescript2';

export const globals = {
  // Apollo
  '@apollo/client/core': 'apollo.client.core',
  '@apollo/client/link/utils': 'apollo.client.link.utils',
  '@apollo/client/link/batch': 'apollo.client.link.batch',

  // GraphQL
  'graphql/language/visitor': 'graphql.visitor',
  'graphql/language/printer': 'graphql.printer',

  // TypeScript
  'tslib': 'tslib',

  // Other
  'zen-observable-ts': 'apolloLink.zenObservable',
  'zen-observable': 'Observable',
};

export default {
  input: 'src/main.ts',
  output: {
    file: 'dist/main.js',
    format: 'cjs',
    exports: 'named',
    globals,
  },
  external: Object.keys(globals),
  plugins: [
    resolve({ module: true }),
    commonjs(),
    typescript({
      tsconfigOverride: {
        compilerOptions: {
          module: 'es2015',
        },
      },
    }),
  ],
};


src/main.ts
import {
  ApolloLink,
  Operation,
  FetchResult,
  serializeFetchParameter,
  selectURI,
  checkFetcher,
  selectHttpOptionsAndBody,
  createSignalIfSupported,
  fallbackHttpConfig,
  HttpOptions,
  ServerParseError,
  Observable,
  fromError,
  throwServerError,
} from '@apollo/client/core';
import { createOperation } from '@apollo/client/link/utils';
import { BatchLink } from '@apollo/client/link/batch';
import combineQuery from 'graphql-combine-query';

/**
 * Transforms Operation for into HTTP results.
 * context can include the headers property, which will be passed to the fetch function
 */
export class AliasBatchHttpLink extends ApolloLink {
  private batchDebounce?: boolean;
  private batchInterval: number;
  private batchMax: number;
  private batcher: ApolloLink;

  constructor(fetchParams?: AliasBatchHttpLink.Options) {
    super();

    let {
      uri = '/graphql',
      // use default global fetch if nothing is passed in
      fetch: fetcher,
      includeExtensions,
      batchInterval,
      batchDebounce,
      batchMax,
      batchKey,
      ...requestOptions
    } = fetchParams || ({} as AliasBatchHttpLink.Options);

    // dev warnings to ensure fetch is present
    checkFetcher(fetcher);

    // fetcher is set here rather than the destructuring to ensure fetch is
    // declared before referencing it. Reference in the destructuring would cause
    // a ReferenceError
    if (!fetcher) {
      fetcher = fetch;
    }

    const linkConfig = {
      http: { includeExtensions },
      options: requestOptions.fetchOptions,
      credentials: requestOptions.credentials,
      headers: requestOptions.headers,
    };

    this.batchDebounce = batchDebounce;
    this.batchInterval = batchInterval || 10;
    this.batchMax = batchMax || 10;

    const batchHandler = (operations: Operation[]) => {
      const chosenURI = selectURI(operations[0], uri);

      const context = operations[0].getContext();

      let combiner: any = combineQuery('__compose');
      operations.forEach((operation, index) => {
        const newArr = [];
        newArr[index] = { ...operation.variables };
        combiner = combiner.addN(operation.query, newArr);
      });
      const { document, variables } = combiner;

      const clientAwarenessHeaders = {};
      if (context.clientAwareness) {
        const { name, version } = context.clientAwareness;
        if (name) {
          clientAwarenessHeaders['apollographql-client-name'] = name;
        }
        if (version) {
          clientAwarenessHeaders['apollographql-client-version'] = version;
        }
      }

      const contextConfig = {
        http: context.http,
        options: context.fetchOptions,
        credentials: context.credentials,
        headers: { ...clientAwarenessHeaders, ...context.headers },
      };

      const newOperation = createOperation(
          {},
          { query: document, operationName: '__compose', variables, context },
      );

      const loadedBody = selectHttpOptionsAndBody(
          newOperation,
          fallbackHttpConfig,
          linkConfig,
          contextConfig,
      );
      const options = loadedBody.options;

      // There's no spec for using GET with batches.
      if (options.method === 'GET') {
        return fromError<FetchResult[]>(new Error('apollo-link-batch-alias does not support GET requests'));
      }

      try {
        (options as any).body = serializeFetchParameter(loadedBody.body, 'Payload');
      } catch (parseError) {
        return fromError<FetchResult[]>(parseError);
      }

      let controller;
      if (!(options as any).signal) {
        const { controller: _controller, signal } = createSignalIfSupported();
        controller = _controller;
        if (controller) (options as any).signal = signal;
      }

      return new Observable<FetchResult[]>((observer) => {
        fetcher(chosenURI, options)
            .then((response) => {
            // Make the raw response available in the context.
              operations.forEach((operation) => operation.setContext({ response }));
              return response;
            })
            .then(parseAndCheckHttpResponse(operations))
            .then(parseCombinedResult(operations))
            .then((result) => {
            // we have data and can send it to back up the link chain
              observer.next(result);
              observer.complete();
              return result;
            })
            .catch((err) => {
              if (err.name === 'AbortError') return;
              if (err.result && err.result.errors && err.result.data) {
                observer.next(err.result);
              }

              observer.error(err);
            });

        return () => {
          // XXX support canceling this request
          // https://developers.google.com/web/updates/2017/09/abortable-fetch
          if (controller) controller.abort();
        };
      });
    };

    batchKey =
      batchKey ||
      ((operation: Operation) => {
        const context = operation.getContext();

        const contextConfig = {
          http: context.http,
          options: context.fetchOptions,
          credentials: context.credentials,
          headers: context.headers,
        };

        // may throw error if config not serializable
        return selectURI(operation, uri) + JSON.stringify(contextConfig);
      });

    this.batcher = new BatchLink({
      batchDebounce: this.batchDebounce,
      batchInterval: this.batchInterval,
      batchMax: this.batchMax,
      batchKey,
      batchHandler,
    });
  }

  request(operation: Operation): Observable<FetchResult> | null {
    return this.batcher.request(operation);
  }
}

export const parseAndCheckHttpResponse = (operations) => (response: Response) => {
  return (
    response
        .text()
        .then((bodyText) => {
          try {
            return JSON.parse(bodyText);
          } catch (err) {
            const parseError = err as ServerParseError;
            parseError.name = 'ServerParseError';
            parseError.response = response;
            parseError.statusCode = response.status;
            parseError.bodyText = bodyText;
            return Promise.reject(parseError);
          }
        })
    // TODO: when conditional types come out then result should be T extends Array ? Array<FetchResult> : FetchResult
        .then((result: any) => {
          if (response.status >= 300) {
          // Network error
            throwServerError(
                response,
                result,
                `Response not successful: Received status code ${response.status}`,
            );
          }
          // TODO should really error per response in a Batch based on properties
          //    - could be done in a validation link
          if (!result.hasOwnProperty('data') && !result.hasOwnProperty('errors')) {
          // Data error
            throwServerError(
                response,
                result,
                `Server response was missing for query '${
              Array.isArray(operations) ? operations.map((op) => op.operationName) : operations.operationName
                }'.`,
            );
          }
          return result;
        })
  );
};

export const parseCombinedResult = (operations: Operation[]) => (result) => {
  return operations.map((op, index) => {
    const selections = op.query.definitions.map((def: any) => def.selectionSet.selections).flat();

    const fields = selections.reduce((acc, s) => {
      const name: string = s.alias?.value || s.name.value;
      acc[name] = result.data[`${name}_${index}`];
      return acc;
    }, {});

    const errors = selections
        .map((s) => {
          const name: string = s.alias?.value || s.name.value;
          return result.errors?.filter((err) => err.path[0] === `${name}_${index}`) || [];
        })
        .flat();

    return { data: fields, errors: errors.length ? errors : null };
  });
};

export namespace AliasBatchHttpLink {
  export type Options = Pick<
    BatchLink.Options,
    'batchMax' | 'batchDebounce' | 'batchInterval' | 'batchKey'
  > & HttpOptions;
}
